3 failing test on non-amd64 platforms (precision-related)

Debian performs regular CI tests on amd64, arm64, and ppc64el (PowerPC) architectures. We've seen three tests fail on the latter two architectures; analysis shows that the failure stems from a slight deviation of actual to the expected value.

I've worked around this in Debian by increasing their relative tolerances (see [patch](https://salsa.debian.org/science-team/tpot/-/blob/debian/master/debian/patches/Adjust-tolerances-on-non-amd64.patch)), but I wanted to report this, in case you see this as something worth investigating.

Specifically, these are the failures (note the actual and expected values):
```
======================================================================
FAIL: Assert that the StackingEstimator worked as expected in scikit-learn pipeline in regression.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/ckk/tpot-0.11.1+dfsg2/tests/stacking_estimator_tests.py", line 113, in test_StackingEstimator_4
    assert np.allclose(known_cv_score, cv_score)
AssertionError: 
>>  assert <module 'numpy' from '/usr/lib/python3/dist-packages/numpy/__init__.py'>.allclose(0.8216045257587923, 0.8207525232725118)
    

======================================================================
FAIL: Assert that the TPOTRegressor score function outputs a known score for a fixed pipeline.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/ckk/tpot-0.11.1+dfsg2/tests/tpot_tests.py", line 580, in test_score_3
    assert np.allclose(known_score, score)
AssertionError: 
-------------------- >> begin captured stdout << ---------------------
Warning: xgboost.XGBRegressor is not available and will not be used by TPOT.

--------------------- >> end captured stdout << ----------------------
>>  assert <module 'numpy' from '/usr/lib/python3/dist-packages/numpy/__init__.py'>.allclose(-11.708199875921563, -11.96209223601317)
    

======================================================================
FAIL: Assert that the TPOTRegressor score function outputs a known score for a fixed pipeline with sample weights.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/ckk/tpot-0.11.1+dfsg2/tests/tpot_tests.py", line 633, in test_sample_weight_func
    assert np.allclose(known_score, score)
AssertionError: 
-------------------- >> begin captured stdout << ---------------------
Warning: xgboost.XGBRegressor is not available and will not be used by TPOT.

--------------------- >> end captured stdout << ----------------------
>>  assert <module 'numpy' from '/usr/lib/python3/dist-packages/numpy/__init__.py'>.allclose(-11.586816877933911, -11.583413861528234)
```

Two are within 1% of the expected value, the third is within 3%.

Would you consider these tolerances acceptable, or are these tests that should absolutely produce the expected value, and nothing else?

--- a/tests/stacking_estimator_tests.py
+++ b/tests/stacking_estimator_tests.py
@@ -23,6 +23,8 @@ License along with TPOT. If not, see <http://www.gnu.org/licenses/>.
 
 """
 
+import platform
+
 import numpy as np
 from tpot.builtins import StackingEstimator
 from sklearn.linear_model import LogisticRegression, Lasso
@@ -103,4 +105,10 @@ def test_StackingEstimator_4():
     cv_score = np.mean(cross_val_score(sklearn_pipeline, training_features_r, training_target_r, cv=3, scoring='r2'))
     known_cv_score = 0.8216045257587923
 
-    assert np.allclose(known_cv_score, cv_score)
+    # On some non-amd64 systems such as arm64, a resulting score of
+    # 0.8207525232725118 was observed, so we need to add a tolerance there
+    if platform.machine() != 'x86_64':
+        assert np.allclose(known_cv_score, cv_score, rtol=0.01)
+    else:
+        assert np.allclose(known_cv_score, cv_score)
+

--- a/tests/tpot_tests.py
+++ b/tests/tpot_tests.py
@@ -52,6 +52,7 @@ from datetime import datetime
 from time import sleep
 from tempfile import mkdtemp
 from shutil import rmtree
+import platform
 
 from sklearn.datasets import load_digits, load_boston, make_classification
 from sklearn.model_selection import train_test_split, cross_val_score, GroupKFold
@@ -573,7 +574,12 @@ def test_score_3():
 
     # Get score from TPOT
     score = tpot_obj.score(testing_features_r, testing_target_r)
-    assert np.allclose(known_score, score)
+    # On some non-amd64 systems such as arm64, a resulting score of
+    # 0.8207525232725118 was observed, so we need to add a tolerance there
+    if platform.machine() != 'amd64':
+        assert np.allclose(known_score, score, rtol=0.03)
+    else:
+        assert np.allclose(known_score, score)
 
 
 def test_sample_weight_func():
@@ -621,7 +627,12 @@ def test_sample_weight_func():
 
     assert np.allclose(cv_score1, cv_score2)
     assert not np.allclose(cv_score1, cv_score_weight)
-    assert np.allclose(known_score, score)
+    # On some non-amd64 systems such as arm64, a resulting score of
+    # 0.8207525232725118 was observed, so we need to add a tolerance there
+    if platform.machine() != 'amd64':
+        assert np.allclose(known_score, score, rtol=0.01)
+    else:
+        assert np.allclose(known_score, score)
 
 
 def test_template_1():

